pkg/network/transport/transport.go

```diff
@@ -0,0 +1,306 @@
+	}
+	if config.CertValidator == nil {
+		return nil, fmt.Errorf("certificate validator required")
+	}
+	if config.Handler == nil {
+		return nil, fmt.Errorf("connection handler required")
+	}
+
+	// Verify the certificate
+	if err := config.CertValidator.ValidateCertificate(config.TLSCert.Leaf); err != nil {
+		return nil, ErrInvalidCertificate
+	}
+
+	return &Transport{
+		config: config,
+		conns:  make(map[string]*Conn),
+	}, nil
+}
+
+// Start initializes the transport listener and begins accepting connections.
+// Returns an error if starting the listener fails.
+func (t *Transport) Start() error {
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{*t.config.TLSCert},
+		NextProtos:         t.config.Handler.GetProtocols(),
+		ClientAuth:         tls.RequireAnyClientCert,
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true,
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			return t.config.CertValidator.ValidateCertificate(c)
+		},
+	}
+
+	listener, err := quic.ListenAddr(t.config.ListenAddr, tlsConfig, &quic.Config{
+		EnableDatagrams: true,
+		MaxIdleTimeout:  MaxIdleTimeout,
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrListenerFailed, err)
+	}
+
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	t.listener = listener
+	t.acceptLoopWg.Add(1)
+	go func() {
+		defer t.acceptLoopWg.Done()
+		t.acceptLoop()
+	}()
+	return nil
+}
+
+// Stop gracefully shuts down the transport and all active connections.
+// Waits for the accept loop to finish before returning.
+func (t *Transport) Stop() error {
+	t.cancel()
+	if t.listener != nil {
+		t.listener.Close()
+	}
+	t.acceptLoopWg.Wait() // Wait for acceptLoop to finish
+	return nil
+}
+
+// Connect initiates a connection to a remote peer.
+// Returns the new connection or an error if connection fails.
+func (t *Transport) Connect(addr string) (*Conn, error) {
+	tlsConf := &tls.Config{
+		Certificates:       []tls.Certificate{*t.config.TLSCert},
+		NextProtos:         t.config.Handler.GetProtocols(),
+		ClientAuth:         tls.RequireAnyClientCert,
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true,
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			if err := t.config.CertValidator.ValidateCertificate(c); err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			return nil
+		},
+	}
+
+	quicConn, err := quic.DialAddr(t.ctx, addr, tlsConf, &quic.Config{
+		EnableDatagrams: true,
+		MaxIdleTimeout:  MaxIdleTimeout,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrDialFailed, err)
+	}
+
+	conn := t.handleConnection(quicConn)
```

Comment:
I think this might be validating the cert twice? The `VerifyPeerCertificate` callback already covers the validation part in `handleConnection` i think.